src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp

```diff
@@ -58,38 +58,178 @@ Archive, Report 2020/1118, 2020. https://eprint.iacr.org/2020/
 
 namespace lbcrypto {
 
+uint32_t ParameterGenerationBGVRNS::computeRingDimension(
+  std::shared_ptr<CryptoParametersBase<DCRTPoly>> cryptoParams,
+  uint32_t qBound, usint cyclOrder) const {
+  const auto cryptoParamsBGVRNS =
+    std::static_pointer_cast<CryptoParametersBGVRNS>(cryptoParams);
+    //// HE Standards compliance logic/check
+  SecurityLevel stdLevel = cryptoParamsBGVRNS->GetStdLevel();
+
+  uint32_t ringDimension = cyclOrder / 2;
+
+  // RLWE security constraint
+  DistributionType distType =
+      (cryptoParamsBGVRNS->GetMode() == RLWE) ? HEStd_error : HEStd_ternary;
+  auto nRLWE = [&](usint q) -> uint32_t {
+    return StdLatticeParm::FindRingDim(distType, stdLevel, q);
+  };
+
+  // Case 1: SecurityLevel specified as HEStd_NotSet -> Do nothing
+  if (stdLevel != HEStd_NotSet) {
+    if (ringDimension == 0) {
+      // Case 2: SecurityLevel specified, but ring dimension not specified
+      // Choose ring dimension based on security standards
+      ringDimension = nRLWE(qBound);
+    } else {
+      // Case 3: Both SecurityLevel and ring dimension specified
+      // Check whether particular selection is standards-compliant
+      auto he_std_n = nRLWE(qBound);
+      if (he_std_n > ringDimension) {
+        OPENFHE_THROW(
+            math_error,
+            "The specified ring dimension (" + std::to_string(ringDimension) +
+                ") does not comply with HE standards recommendation (" +
+                std::to_string(he_std_n) + ").");
+      }
+    }
+  } else if (ringDimension == 0) {
+    OPENFHE_THROW(
+        math_error,
+        "Please specify the ring dimension or desired security level.");
+  }
+  return ringDimension;
+}
+
+std::pair<std::vector<NativeInteger>, uint32_t> ParameterGenerationBGVRNS::computeModuli(
+  std::shared_ptr<CryptoParametersBase<DCRTPoly>> cryptoParams,
+  uint32_t ringDimension,
+  int32_t evalAddCount, int32_t keySwitchCount,
+  usint relinWindow, usint firstModSize, uint32_t auxBits,
+  enum KeySwitchTechnique ksTech,
+  usint numPrimes) const {
+
+  std::vector<NativeInteger> moduliQ(numPrimes + 1);
+  
+  const auto cryptoParamsBGVRNS =
+      std::static_pointer_cast<CryptoParametersBGVRNS>(cryptoParams);
+  double sigma = cryptoParamsBGVRNS->GetDistributionParameter();
+  double alpha = cryptoParamsBGVRNS->GetAssuranceMeasure();
+  double plainModulus = static_cast<double>(cryptoParamsBGVRNS->GetPlaintextModulus());
+
+  // Bound of the Gaussian error polynomial
+  double Berr = sigma * sqrt(alpha);
+  // Bound of the key polynomial
+  double Bkey;
```

Comment:
uninitialized variable. ternary operator should be used here

Reply:
Fixed

Reply:
Fixed